arrow-libs/fx/arrow-collectors/src/commonMain/kotlin/arrow/collectors/Collectors.kt

```diff
@@ -0,0 +1,147 @@
+package arrow.collectors
+
+import arrow.atomic.Atomic
+import arrow.atomic.AtomicInt
+import arrow.atomic.update
+
+/**
+ * Library of [Collector]s.
+ */
+public object Collectors {
+
+  /**
+   * Returns always the same value, regardless of the collected flow.
+   *
+   * @param value Value returns as result of the collection.
+   */
+  public fun <R> constant(value: R): Collector<Any?, R> =
+    object : CollectorI<R, Any?, R> {
+      override val characteristics: Set<Characteristics> = Characteristics.IDENTITY_CONCURRENT_UNORDERED
+      override suspend fun supply(): R = value
+      override suspend fun accumulate(current: R, value: Any?) {}
+      override suspend fun finish(current: R): R = current
+    }
+
+  /**
+   * Counts the number of elements in the flow.
+   */
+  public val length: Collector<Any?, Int> =
+    object : CollectorI<AtomicInt, Any?, Int> {
+      override val characteristics: Set<Characteristics> = Characteristics.CONCURRENT_UNORDERED
+      override suspend fun supply(): AtomicInt = AtomicInt(0)
+      override suspend fun accumulate(current: AtomicInt, value: Any?) {
+        current.incrementAndGet()
+      }
+
+      override suspend fun finish(current: AtomicInt): Int = current.get()
+    }
+
+  /**
+   * Sum of all the values in the flow.
+   */
+  public val sum: Collector<Int, Int> =
+    intReducer({ 0 }, Int::plus)
+
+  public fun intReducer(
+    initial: () -> Int, combine: (Int, Int) -> Int,
+  ): Collector<Int, Int> =
+    object : CollectorI<AtomicInt, Int, Int> {
+      override val characteristics: Set<Characteristics> = emptySet()
+      override suspend fun supply(): AtomicInt = AtomicInt(initial())
+      override suspend fun accumulate(current: AtomicInt, value: Int) {
+        current.update { combine(it, value) }
+      }
+
+      override suspend fun finish(current: AtomicInt): Int = current.get()
+    }
+
+  public fun <M> reducer(
+    initial: () -> M, combine: (M, M) -> M, unordered: Boolean = false,
+  ): Collector<M, M> =
+    object : CollectorI<Atomic<M>, M, M> {
+      override val characteristics: Set<Characteristics> =
+        if (unordered) Characteristics.CONCURRENT_UNORDERED
+        else emptySet()
+
+      override suspend fun supply(): Atomic<M> = Atomic(initial())
+      override suspend fun accumulate(current: Atomic<M>, value: M) {
+        current.update { combine(it, value) }
+      }
+
+      override suspend fun finish(current: Atomic<M>): M = current.get()
+    }
+
+  /**
+   * Returns the "best" value from the value.
+   *
+   * @param selectNew Decides whether the new value is "better" than the previous best.
+   */
+  public fun <M> bestBy(
+    selectNew: (old: M, new: M) -> Boolean,
+  ): Collector<M, M?> =
+    object : CollectorI<Atomic<M?>, M, M?> {
```

Comment:
If `CollectorI` is used to build operators and it's something that would be part of what users can do perhaps we should include `@JvmOverload` constructors or factory methods in `CollectorI`'s companion so we can do construction of the possible combinations without having to implement the interface in an anonymous object.
```kotlin
public fun <T> list(): Collector<T, List<T>> =
    CollectorI(
        mutableListOf(), 
        setOf(Characteristics.IDENTITY_FINISH), 
        { current: MutableList<T>, value: T -> current.add(value) },
        ::identity
    )
```

Reply:
You're completely right. I've reworked the code in that form, and it looks waaaay better than before!